packages/dd-trace/src/encode/0.4.js

```diff
@@ -1,165 +1,50 @@
-  offset += 8 // the int64 will live here
-
-  offset = state.copy(offset, fields.error)
-  const errorOffset = offset
-  offset = state.copy(offset, tokens.int[0])
-
-  return {
-    headerBuffer: state.buffer.slice(0, offset),
-    traceIdOffset,
-    spanIdOffset,
-    startOffset,
-    durationOffset,
-    errorOffset
+class AgentEncoder {
+  constructor (writer) {
+    this._writer = writer
+    this._reset()
   }
-})()
-
-function encode (initBuffer, offset, trace, initWriter) {
-  state = new EncoderState(initBuffer, offset, trace, initWriter)
-  offset = state.writeArrayPrefix(offset, trace)
-  for (const span of trace) {
-    let fieldCount = 9
-
-    span.parent_id && fieldCount++
-    span.type && fieldCount++
-    span.metrics && fieldCount++
-
-    offset = state.copy(offset, tokens.map[fieldCount])
-
-    offset = copyHeader(offset, span)
-
-    if (span.parent_id) {
-      offset = state.copy(offset, fields.parent_id)
-      offset = state.copy(offset, tokens.uint64)
-      offset += util.writeId(state.buffer, span.parent_id, offset)
-    }
 
-    offset = state.copy(offset, fields.name)
-    offset = write(offset, span.name)
+  count () {
+    return this._traces.length
+  }
 
-    offset = state.copy(offset, fields.resource)
-    offset = write(offset, span.resource)
+  encode (trace) {
+    const buffer = encode(trace)
 
-    offset = state.copy(offset, fields.service)
-    offset = write(offset, span.service)
+    log.debug(() => `Adding encoded trace to buffer: ${buffer.toString('hex').match(/../g).join(' ')}`)
 
-    if (span.type) {
-      offset = state.copy(offset, fields.type)
-      offset = write(offset, span.type)
+    if (buffer.length + this._size > HARD_LIMIT) {
+      this._writer.flush()
     }
 
-    offset = state.copy(offset, fields.meta)
-    offset = state.writeMap(offset, span.meta, write)
-
-    if (span.metrics) {
-      offset = state.copy(offset, fields.metrics)
-      offset = state.writeMap(offset, span.metrics, write)
-    }
+    this._size += buffer.length
+    this._traces.push(buffer)
   }
 
-  return offset
-}
+  makePayload () {
+    const prefix = Buffer.alloc(5)
 
-function copyHeader (offset, span) {
-  util.writeId(headerBuffer, span.trace_id, traceIdOffset)
-  util.writeId(headerBuffer, span.span_id, spanIdOffset)
-  util.writeInt64(headerBuffer, span.start, startOffset)
-  util.writeInt64(headerBuffer, span.duration, durationOffset)
-  headerBuffer.set(tokens.int[span.error], errorOffset)
-  return state.copy(offset, headerBuffer)
-}
+    prefix[0] = 0xdd
+    prefix.writeUInt32BE(this._traces.length, 1)
 
-function write (offset, val) {
-  if (typeof val === 'string') {
-    return state.copy(offset, cachedString(val))
-  } else { // val is number
-    offset = state.checkOffset(offset, 9)
-    state.buffer.writeUInt8(0xcb, offset)
-    state.buffer.writeDoubleBE(val, offset + 1)
-    return offset + 9
-  }
-}
+    const buffer = [prefix].concat(this._traces)
```

Comment:
Pretty much every time you have a case like this, you probably actually want to do:

`[prefix, ...this._traces]`

This applies everywhere else in this PR, too.